File: src/main/java/com/mrkirby153/playtimetracker/command/CommandPlayTime.java
package com.mrkirby153.playtimetracker.command;

import com.mojang.authlib.GameProfile;
import com.mojang.brigadier.CommandDispatcher;
import com.mrkirby153.playtimetracker.repository.PlaySession;
import com.mrkirby153.playtimetracker.repository.PlaytimeData;
import com.mrkirby153.playtimetracker.repository.PlaytimeRepository;
import me.mrkirby153.kcutils.Time;
import net.minecraft.command.CommandSource;
import net.minecraft.command.Commands;
import net.minecraft.command.ISuggestionProvider;
import net.minecraft.command.arguments.GameProfileArgument;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.text.StringTextComponent;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

public class CommandPlayTime {

    private static final Logger LOGGER = LogManager.getLogger();

    public static void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
            Commands.literal("playtime").requires((source) -> source.hasPermission(0))
                .executes(sender -> {
                    CommandSource source = sender.getSource();
                    if (source.getEntity() instanceof PlayerEntity) {
                        PlayerEntity entity = (PlayerEntity) source.getEntity();
                        return showPlaytime(source, entity.getGameProfile());
                    } else {
                        return showLeaderboard(source);
                    }
                })
                .then(Commands.literal("get").then(
                    Commands.argument("target", GameProfileArgument.gameProfile())
                        .suggests((sender, suggest) -> {
                            PlaytimeData data = PlaytimeRepository.loadFromServer(
                                sender.getSource().getServer());
                            return ISuggestionProvider.suggest(data.getAllUsernames(), suggest);
                        }).executes(sender -> {
                            try {
                                List<GameProfile> profiles = new ArrayList<>(
                                    GameProfileArgument.getGameProfiles(sender, "target"));
                                GameProfile p = profiles.get(0);
                                return showPlaytime(sender.getSource(), p);
                            } catch (Exception e) {
                                sender.getSource()
                                    .sendFailure(new StringTextComponent("Failed: " + e.getMessage()));
                                return 0;
                            }
                        })
                ))
                .then(Commands.literal("list")
                    .executes(sender -> showLeaderboard(sender.getSource()))));
    }


    private static <K, V extends Comparable<V>> LinkedHashMap<K, V> sortByValue(Map<K, V> map) {
        LinkedList<Map.Entry<K, V>> list = new LinkedList<>(map.entrySet());
        list.sort((o1, o2) -> o2.getValue().compareTo(o1.getValue()));

        LinkedHashMap<K, V> sorted = new LinkedHashMap<>();
        list.forEach(e -> {
            sorted.put(e.getKey(), e.getValue());
        });
        return sorted;
    }

    private static int showPlaytime(CommandSource source, GameProfile profile) {
        List<PlaySession> sessions = PlaytimeRepository.loadFromServer(
            source.getServer()).get(profile.getId());
        long totalPlayTime = sessions.stream().map(PlaySession::duration)
            .reduce(0L, Long::sum);
        float avg = totalPlayTime / (float) sessions.size();

        source.sendSuccess(new StringTextComponent(
            String.format("Play time for %s", profile.getName())), false);
        source.sendSuccess(new StringTextComponent(
            String.format("Time Played: %s over %d sessions",
                Time.formatLong(totalPlayTime), sessions.size())), false);
        source.sendSuccess(new StringTextComponent(
            String.format("Average Time Played: %s", Time.formatLong(
                (long) avg))), false);
        long lastLogout = 0;
        boolean active = false;
        PlaySession currentSession = null;
        for (PlaySession s : sessions) {
            if (s.isActive()) {
                active = true;
                currentSession = s;
                break;
            }
            if (s.getEnd() > lastLogout) {
                lastLogout = s.getEnd();
            }
        }
        String lastSeen = active ? "Online"
            : String.format("%s ago", Time.formatLong(System.currentTimeMillis() - lastLogout));
        if(active) {
            source.sendSuccess(new StringTextComponent(String.format("Current Session: %s", Time.formatLong(currentSession.duration()))), false);
        } else {
            source.sendSuccess(new StringTextComponent(String.format("Last Seen: %s", lastSeen)),
                false);
        }
        return 1;
    }

    private static int showLeaderboard(CommandSource source) {
        Map<UUID, Long> playtime = new HashMap<>();
        PlaytimeData playtimeData = PlaytimeRepository.loadFromServer(
            source.getServer());
        playtimeData.getAll()
            .forEach((uuid, playSessions) -> {
                playtime.put(uuid,
                    playSessions.stream().map(PlaySession::duration)
                        .reduce(0L, Long::sum));
            });

        LinkedHashMap<UUID, Long> sorted = sortByValue(playtime);

        AtomicInteger num = new AtomicInteger(1);
        boolean displayed = false;
        Entity sourceEntity = source.getEntity();
        for (Map.Entry<UUID, Long> entry : sorted.entrySet()) {
            if (num.get() > 5 && sourceEntity instanceof PlayerEntity) {
                break;
            }
            if (sourceEntity instanceof PlayerEntity) {
                LOGGER.debug("Player: {} {}", entry.getKey(), sourceEntity.getUUID());
                if (entry.getKey().equals(sourceEntity.getUUID())) {
                    LOGGER.debug("Displayed!");
                    displayed = true;
                }
            }
            source.sendSuccess(
                new StringTextComponent(
                    String.format("#%d: %s - %s", num.getAndIncrement(),
                        playtimeData.getUsername(entry.getKey()),
                        Time.formatLong(entry.getValue()))), false);
        }

        if (sourceEntity instanceof PlayerEntity && !displayed) {
            int place = 1;
            for (Map.Entry<UUID, Long> e : sorted.entrySet()) {
                if (e.getKey() == sourceEntity.getUUID()) {
                    break;
                }
                place++;
            }
            source.sendSuccess(new StringTextComponent(""), false);
            long time = sorted.get(sourceEntity.getUUID());
            source.sendSuccess(new StringTextComponent(
                    String.format("#%d: %s - %s", place,
                        sourceEntity.getName().getString(), Time.formatLong(time))),
                false);
        }
        return 1;
    }
}

File: src/main/kotlin/com/mrkirby153/playtime/command/CommandPlayTime.kt
package com.mrkirby153.playtime.command

import com.mrkirby153.playtime.PlaytimeTracker
import me.mrkirby153.kcutils.Time
import net.minecraft.command.CommandBase
import net.minecraft.command.ICommandSender
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.server.MinecraftServer
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.Style
import net.minecraft.util.text.TextComponentString
import net.minecraft.util.text.TextFormatting
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Comparator
import java.util.LinkedList
import java.util.Locale
import java.util.UUID

class CommandPlayTime : CommandBase() {
    override fun getName(): String {
        return "playtime"
    }

    override fun execute(server: MinecraftServer, sender: ICommandSender, args: Array<out String>) {
        if (args.isEmpty()) {
            // Display the play times for everyone
            val playTimes = mutableMapOf<UUID, Long>()
            PlaytimeTracker.instance.repository.repo.forEach { (key, value) ->
                playTimes[key] = value.getTotalPlaytime()
            }

            val sorted = sortByValue(playTimes)
            var num = 1

            var displayed = false

            sorted.forEach { k, v ->
                if (num > 5 && sender is EntityPlayer)
                    return@forEach
                if (sender is EntityPlayer) {
                    if (k == sender.uniqueID)
                        displayed = true
                }
                sender.sendMessage(TextComponentString("#${num++}: ${PlaytimeTracker.instance.usernameRepo.getName(k)} - ${Time.formatLong(v)}"))
            }
            var place = 1
            sorted.forEach { k, _ ->
                if (sender is EntityPlayer) {
                    if (sender.uniqueID == k)
                        return@forEach
                    else
                        place++
                }
            }
            if (sender is EntityPlayer && !displayed) {
                sender.sendMessage(TextComponentString(""))
                val time = playTimes[sender.uniqueID]
                sender.sendMessage(TextComponentString("#$place: ${sender.name} = ${if (time != null) Time.formatLong(time) else "Unknown"}"))
            }
        } else if (args.size == 1) {
            val user = args[0]
            val uuid = PlaytimeTracker.instance.usernameRepo.getUUID(user)
            if (uuid == null) {
                sender.sendMessage(TextComponentString("That user does not exist!").apply {
                    style = Style().apply { color = TextFormatting.RED }
                })
                return
            }

            val playTime = PlaytimeTracker.instance.repository.repo[uuid] ?: return

            val sessions = playTime.sessions

            var sessionCount = sessions.size

            var average = 0.0
            sessions.forEach {
                average += it.duration
            }
            if (playTime.currentSession != null) {
                sessionCount++
                average += playTime.currentSession!!.duration
            }
            average /= sessionCount

            sender.sendMessage(TextComponentString("Playtime for $user").apply {
                style = Style().apply { color = TextFormatting.GREEN }
            })

            sender.sendMessage(TextComponentString(
                    "Time Played: ${Time.formatLong(playTime.getTotalPlaytime())} over $sessionCount sessions"))

            sender.sendMessage(TextComponentString("Current Session: ${Time.formatLong(playTime.getCurrentPlaytime())}"))
            sender.sendMessage(TextComponentString("Average Time Played: ${Time.formatLong(average.toLong())}"))
        }
    }

    override fun getUsage(sender: ICommandSender?): String {
        return "/playtime"
    }

    fun formatDouble(double: Double): Double = DecimalFormat("#.##", DecimalFormatSymbols(Locale.US)).format(double).toDouble()

    fun <K, V : Comparable<V>> sortByValue(map: Map<K, V>): Map<K, V> {
        val list = LinkedList<Map.Entry<K, V>>(map.entries)

        list.sortWith(Comparator { o1, o2 ->
            o2.value.compareTo(o1.value)
        })

        val sortedMap = LinkedHashMap<K, V>()

        list.forEach {
            sortedMap.put(it.key, it.value)
        }
        return sortedMap
    }

    override fun checkPermission(server: MinecraftServer?, sender: ICommandSender?): Boolean {
        return true
    }

    override fun getTabCompletions(server: MinecraftServer, sender: ICommandSender,
                                   args: Array<String>, targetPos: BlockPos?): MutableList<String> {
        return if(args.isNotEmpty()) getListOfStringsMatchingLastWord(args, PlaytimeTracker.instance.usernameRepo.repo.values.toList()) else mutableListOf()
    }
}